## COMM 313 Final Project - Blog Post / Data Story 
### Jessica Pasik

## Additional Modules

In [2]:
#Additional modules
import os
import pandas as pd
import re
import json
import requests
from bs4 import BeautifulSoup
import lyricsgenius
from collections import Counter
import nltk
from nltk import Text
nltk.download('stopwords')
from nltk.corpus import stopwords
stop_words = stopwords.words('english')
sect_stoppers = ['pre-chorus','refrain','chorus','verse','intro','outro','bridge','verse 1','verse 2','verse 3','verse 4','1','2','3','4','Tim McGraw','Faith Hill','Tim McGraw & Faith Hill']
for x in sect_stoppers:
    stop_words.append(x)
# pos tagging
from nltk import pos_tag, pos_tag_sents, FreqDist, ConditionalFreqDist

In [4]:
%run '../data_analysis/functions.ipynb'

In [12]:
%run '../data_analysis/frequency_ngram_analysis.ipynb'

Top 50 words in your `all_90s` corpus
[('love', 108), ("i'm", 75), ('know', 74), ('oh', 70), ('yeah', 56), ('like', 55), ('got', 50), ('get', 47), ('let', 47), ('little', 45), ('wanna', 44), ("ain't", 42), ('one', 38), ('never', 38), ('way', 37), ('tell', 36), ('girl', 35), ("i've", 35), ('take', 34), ('ya', 34), ('come', 33), ('say', 33), ('go', 33), ('boy', 32), ('said', 32), ('make', 32), ('baby', 30), ('heart', 30), ('away', 29), ('gonna', 28), ('think', 28), ('well', 27), ('right', 27), ('maria', 27), ('time', 26), ('man', 25), ('ever', 25), ('feel', 25), ('want', 24), ('night', 22), ('would', 22), ('knows', 22), ('world', 21), ('hey', 21), ("can't", 21), ('kiss', 21), ('maybe', 19), ('really', 19), ('passionate', 19), ('kisses', 19)]
Top 50 # of songs each type occurs in your `all_90s` corpus
[('know', 22), ("i'm", 21), ('like', 20), ('love', 19), ('oh', 19), ('night', 19), ('got', 18), ('one', 18), ('said', 17), ('never', 17), ('get', 16), ("ain't", 15), ('right', 15), ("i've", 

Top 50 words in your `all_male` corpus
[("i'm", 114), ('like', 98), ('love', 78), ('know', 70), ('yeah', 60), ('got', 57), ('wanna', 57), ('get', 55), ("ain't", 55), ('oh', 51), ('baby', 51), ('ya', 48), ('back', 46), ('make', 44), ('little', 43), ('girl', 40), ('never', 39), ('think', 38), ('take', 37), ('right', 37), ('oooh', 36), ('one', 35), ('see', 34), ('heart', 33), ('go', 33), ('gonna', 33), ('come', 32), ('tell', 31), ('rock', 30), ('night', 29), ("can't", 29), ("i've", 28), ('way', 28), ('mama', 28), ("'em", 28), ('need', 28), ('maria', 27), ('hey', 26), ('say', 25), ('man', 25), ('around', 25), ("i'll", 24), ('whiskey', 23), ('world', 23), ('beautiful', 23), ('let', 22), ('away', 21), ('time', 21), ('good', 21), ('would', 20)]
Top 50 # of songs each type occurs in your `all_male` corpus
[('know', 24), ("i'm", 24), ('like', 22), ('night', 19), ('get', 19), ('got', 19), ('go', 18), ('never', 17), ('yeah', 17), ("ain't", 17), ('right', 16), ('love', 15), ('oh', 14), ('take', 14

Top 50 words in your `male_2010s` corpus
[('like', 72), ("i'm", 57), ('back', 44), ('baby', 40), ('oooh', 36), ('yeah', 31), ('need', 28), ('rock', 28), ('got', 27), ('right', 27), ("ain't", 27), ('little', 26), ("'em", 26), ('mama', 26), ('make', 25), ('think', 25), ('know', 24), ('get', 24), ('gonna', 23), ('go', 23), ('wanna', 23), ('see', 21), ('hey', 20), ('dirt', 19), ('one', 19), ('good', 18), ('way', 17), ('whiskey', 17), ('used', 16), ('road', 16), ("i'ma", 16), ('tequila', 16), ('around', 15), ('night', 15), ('hell', 15), ('free', 15), ('drink', 14), ('take', 14), ("'cause", 14), ('man', 14), ('ya', 14), ('crazy', 14), ('always', 14), ('country', 13), ('never', 13), ('glasses', 13), ('drunk', 13), ("can't", 12), ('feel', 12), ('shine', 12)]
Top 50 # of songs each type occurs in your `male_2010s` corpus
[('like', 12), ('go', 12), ('back', 12), ('know', 11), ("i'm", 11), ('yeah', 11), ('get', 11), ("ain't", 10), ('got', 9), ('baby', 9), ('right', 8), ('take', 8), ('way', 8), ('

In [38]:
%run '../data_analysis/partofspeech_tagging_analysis.ipynb'

39.8196270253745 percent of the tokens in all 2010s songs were verbs
16.70742892081932 percent of the tokens in all 2010s songs were adjectives
42.479946524064175 percent of the tokens in all 90s songs were verbs
17.34625668449198 percent of the tokens in all 90s songs were adjectives
43.145624692168774 percent of the tokens in all female songs were verbs
17.271384009193895 percent of the tokens in all female songs were adjectives
43.145624692168774 percent of the tokens in all male songs were verbs
17.271384009193895 percent of the tokens in all male songs were adjectives


In [5]:
char_to_strip = '.,!][?;$"-()'

In [6]:
all_charts = json.load(open('../data/charts/all_charts.json'))

## What can we learn about the evolution of country music from analyzing lyrics from the 1990s and 2010s by male and female artists? 
## Jessica Pasik
### COMM 313 - Spring 2021

## INTRODUCTION

Despite some people's aversion to country music, the genre remains one of the most-listened to and popular musical subcategories of our generation. The country music genre, dating back to the 1920s, has experienced various different musical tunes and styles; country music started with 'hillbilly' tunes, experimented with bluegrass and gospel music, and was followed by outlaw country and more folk and soft rock tunes. The most recent decade saw the injection of pop and rap into country music songs. Taking into account all of these details and stylistic shifts, I thought it best to dig a little deeper and see what insights I might find when analyzing country music lyrics over time.

When thinking about this project, I knew I had to narrow down my data set to make my research questions and hypotheses more targeted. As a result, I sought out to perform an exploratory analysis of country music lyrics from the 1990s - a big decade for country music - and the 2010s, to see if there were any interesting or significant patterns that stand out that could inform me how the country genre has evolved over the years. 

Keeping this in mind, I have shared my research questions and hypotheses that will be tested in my data analysis below:

### Research Question

How has the country music genre changed over time – from the 1990s to 2010s? Are there any prevalent changes between men and women both within and between these decades? What evidence can I find through lyrical analysis that could tell me more about where the country genre is headed in the future? What can be expected?

### Hypothesis 

I hypothesize that while there may be diversity in topics over the decades, the lyrical analysis will reveal great similarity in lyrics from the 1990s and 2010s; I believe any substantial lyrical differences will be more prominent between male and female artists. 

## COLLECTING, ORGANIZING, & BUILDING THE CORPORA

As I mentioned earlier, in order to narrow my findings, I decided to focus on country songs from the 1990s - a decade full of success and fame for the country genre - and the 2010s. I chose these two time periods with the hopes that I could distinguish lyrical differences between old country styles of the 1990s and more modern country tunes of the 2010s. In addition to the temporal aspect, I also decided to use gender as another unit of analysis so that I could see if there were any differences in the song lyrics written by male and female artists both between and within the decades. 

Given that I could not find a pre-made data set or chart online that included the exact units of analysis I was looking for, I created an Excel chart of top songs from the 1990s and 2010s by male and female artists. I then transformed this chart into a Pandas data frame, which I then turned into separate dictionaries. After this, I relied on Genius API to scrape the lyrics for every song. Once I compiled all my lyrical data, I further separated and organized my data into various JSON files. The main corpus that will be referenced in all data analysis folders is `all_charts`; `all_charts` pulls lyrics from a combined 80 songs (40 per decade; 20 songs per gender per decade).

## FINDINGS

Methods of analysis used in the project:

- Tokenization
- Frequencies
- N-gram Lists
- Word Comparisons
- KWIC Concordance Analysis
- Collocation Analysis 
- Keyness Analysis
- Part-of-Speech Tagging 

Now, to begin the data story!

## Frequencies, N-gram lists, and Word Comparisons

Using a massive loop that incorporated tokenization, frequencies, and n-gram lists, I was able to run such analyses through the various lyrical dictionaries I saved as JSON files.  

For purposes of this data story, I have only included the most relevant dictionaries for analysis in this notebook.

## All 1990s 

In [13]:
print('Top 50 words in your `all_90s` corpus\n', '='*34, sep='')
print(word_freq_90s.most_common(50))

print('Top 50 # of songs each type occurs in your `all_90s` corpus\n', '='*34, sep='')
print(song_freq_90s.most_common(50))

print('Top 50 bigrams in your `all_90s` corpus\n', '='*34, sep='')
print(bigrams_90s_dist.most_common(50))

print('Top 50 trigrams in your `all_90s` corpus\n', '='*34, sep='')
print(trigrams_90s_dist.most_common(50))

Top 50 words in your `all_90s` corpus
[('love', 108), ("i'm", 75), ('know', 74), ('oh', 70), ('yeah', 56), ('like', 55), ('got', 50), ('get', 47), ('let', 47), ('little', 45), ('wanna', 44), ("ain't", 42), ('one', 38), ('never', 38), ('way', 37), ('tell', 36), ('girl', 35), ("i've", 35), ('take', 34), ('ya', 34), ('come', 33), ('say', 33), ('go', 33), ('boy', 32), ('said', 32), ('make', 32), ('baby', 30), ('heart', 30), ('away', 29), ('gonna', 28), ('think', 28), ('well', 27), ('right', 27), ('maria', 27), ('time', 26), ('man', 25), ('ever', 25), ('feel', 25), ('want', 24), ('night', 22), ('would', 22), ('knows', 22), ('world', 21), ('hey', 21), ("can't", 21), ('kiss', 21), ('maybe', 19), ('really', 19), ('passionate', 19), ('kisses', 19)]
Top 50 # of songs each type occurs in your `all_90s` corpus
[('know', 22), ("i'm", 21), ('like', 20), ('love', 19), ('oh', 19), ('night', 19), ('got', 18), ('one', 18), ('said', 17), ('never', 17), ('get', 16), ("ain't", 15), ('right', 15), ("i've", 

In [14]:
## Calculating type-token ratio for 90s
# Finding amount of unique words across corpora

ttr_90s = []

for song in all_charts['all_90s']:
    toks_90s = song['tokens']
    ttr = len(set(toks_90s)) / len(toks_90s) * 100
    ttr_90s.append(ttr)

print('the average type-token ratio among `all_90s` songs is {}'.format(sum(ttr_90s) / len(ttr_90s)))


the average type-token ratio among `all_90s` songs is 40.612394807300554


## All 2010s

In [15]:
print('Top 50 words in your `all_2010s` corpus\n', '='*34, sep='')
print(word_freq_2010s.most_common(50))

print('Top 50 # of songs each type occurs in your `all_2010s` corpus\n', '='*34, sep='')
print(song_freq_2010s.most_common(50))

print('Top 50 bigrams in your `all_2010s` corpus\n', '='*34, sep='')
print(bigrams_2010s_dist.most_common(50))

print('Top 50 trigrams in your `all_2010s` corpus\n', '='*34, sep='')
print(trigrams_2010s_dist.most_common(50))

Top 50 words in your `all_2010s` corpus
[('like', 108), ("i'm", 97), ('yeah', 87), ('back', 80), ('little', 62), ('get', 60), ('got', 58), ("ain't", 53), ('baby', 52), ('every', 51), ('know', 46), ('never', 45), ('gonna', 43), ('oh', 42), ('take', 40), ("'em", 39), ('make', 38), ('right', 36), ('one', 36), ('oooh', 36), ('see', 35), ('go', 35), ('road', 34), ('good', 33), ('need', 33), ('hope', 33), ('think', 32), ('way', 31), ('rock', 31), ('away', 30), ('ever', 29), ('mama', 29), ('around', 28), ('hey', 28), ('heart', 27), ('name', 27), ('wanna', 27), ('could', 26), ('changed', 25), ("'cause", 24), ('night', 24), ("i'll", 23), ('thing', 22), ('dirt', 22), ('girl', 22), ('song', 22), ('free', 22), ('man', 21), ('come', 21), ("can't", 21)]
Top 50 # of songs each type occurs in your `all_2010s` corpus
[('like', 23), ('back', 22), ('get', 22), ('yeah', 21), ("i'm", 19), ('know', 18), ('go', 18), ('got', 17), ("ain't", 17), ('take', 15), ('one', 15), ('good', 14), ('right', 14), ('baby', 

In [16]:
## Calculating type-token ratio for 2010s
# Finding amount of unique words across corpora

ttr_2010s = []

for song in all_charts['all_2010s']:
    toks_2010s = song['tokens']
    ttr_10s = len(set(toks_2010s)) / len(toks_2010s) * 100
    ttr_2010s.append(ttr_10s)

print('the average type-token ratio among 2010s songs is {}'.format(sum(ttr_2010s) / len(ttr_2010s)))


the average type-token ratio among 2010s songs is 37.27572078056541


In [17]:
## Comparing some key words present in BOTH the 90s and 2010s lyrics
## Key words chosen from printed results earlier in this notebook 

words_to_compare = ['heart','love','baby','girl','yeah','man','little']
print("{:<10} {:<10} {:<10}".format("word", "1990s", "2010s"))
print("="*30)

for word in words_to_compare:
    print("{:<10} {:<10} {:<10}".format(word,
                                   word_freq_90s.get(word,0),
                                   word_freq_2010s.get(word,0)))

word       1990s      2010s     
heart      30         27        
love       108        19        
baby       30         52        
girl       35         22        
yeah       56         87        
man        25         21        
little     45         62        


### Observations - 1990s vs 2010s

When looking at the frequencies and n-grams from the 1990s, there are frequent references to filler words such as "oh", "yeah", and "really", and "hey". However, there are also significant outliers that will be analyzed further throughout this notebook such as "love", "heart", "baby", and "kisses". Additionally, the average type-token ratio - or the amount of unique words across the `all_90s` data is 40.612394807300554, showing there was not significant variety in words used. 

Compared to the data from the 1990s, the frequencies and n-grams from the `all_2010s` data displayed a lot more lyrical variety, with words such as "dirt", "road", "mama" and "hope". There were also some distinct word pairs such as "whiskey glasses", "taste tequila", and "drunk plane" that seem to detail a storyline not present in the 1990s chart. Also, the average type-token ratio among 2010s songs is 37.27572078056541, only slightly lower than that of the 1990s.

Above, I included a chart of words that appeared in the top 50 of both the 1990s and 2010s chart. Words such as "love" and "heart" appeared in both and insinuate storylines about falling in and out of love. I will analyze this further later in this notebook. This finding shows that there may be more similarity in lyrics between the decades than I thought; while the tunes of the songs may have evolved, maybe the lyrical uniqueness has not changed as much.

## All Female

In [20]:
print('Top 50 words in your `all_female` corpus\n', '='*34, sep='')
print(f_word_freq.most_common(50))

Top 50 words in your `all_female` corpus
[('yeah', 83), ('like', 65), ('little', 64), ('oh', 61), ("i'm", 58), ('get', 52), ('got', 51), ('know', 50), ('love', 49), ('back', 49), ('every', 46), ('never', 44), ('let', 43), ("ain't", 40), ('way', 40), ('one', 39), ('away', 38), ('gonna', 38), ('ever', 37), ('take', 37), ('go', 35), ('hope', 35), ('boy', 34), ('said', 34), ('baby', 31), ('well', 27), ('kiss', 27), ('right', 26), ('make', 26), ('time', 25), ('thing', 25), ('feel', 25), ('could', 25), ('changed', 25), ('good', 24), ('prechorus', 24), ('heart', 24), ('say', 23), ('hey', 23), ('come', 22), ('name', 22), ('day', 22), ('put', 22), ('maybe', 22), ('knows', 22), ("i've", 22), ('think', 22), ('road', 21), ('man', 21), ('left', 21)]


## All Male

In [19]:
print('Top 50 words in your `all_male` corpus\n', '='*34, sep='')
print(m_word_freq.most_common(50))

Top 50 words in your `all_male` corpus
[("i'm", 114), ('like', 98), ('love', 78), ('know', 70), ('yeah', 60), ('got', 57), ('wanna', 57), ('get', 55), ("ain't", 55), ('oh', 51), ('baby', 51), ('ya', 48), ('back', 46), ('make', 44), ('little', 43), ('girl', 40), ('never', 39), ('think', 38), ('take', 37), ('right', 37), ('oooh', 36), ('one', 35), ('see', 34), ('heart', 33), ('go', 33), ('gonna', 33), ('come', 32), ('tell', 31), ('rock', 30), ('night', 29), ("can't", 29), ("i've", 28), ('way', 28), ('mama', 28), ("'em", 28), ('need', 28), ('maria', 27), ('hey', 26), ('say', 25), ('man', 25), ('around', 25), ("i'll", 24), ('whiskey', 23), ('world', 23), ('beautiful', 23), ('let', 22), ('away', 21), ('time', 21), ('good', 21), ('would', 20)]


### Observations - Female vs Male 

When looking at the `all_female` chart, it is clear that the words "little", "love", "heart", and "kiss" stand out. Looking at these words without contextualization, there seems to be a common theme of 'love' and love-related topics in country songs written by female artists from both the 1990s and 2010s. Compared to the female chart, the `all_male` chart used the world "love" significantly more; however, it also had much more lyrical variety with references to "whiskey", "world", and "mama". In other words, it is harder to pick up a common theme for songs written by male artists from both decades. Looking at these frequencies, one can see that male and female artists *do* in fact differ (not always in a big way) in the topics and subjects they decide to sing about, thus supporting my hypothesis that there would be some lyrical distinction across genders. 

## KWIC Concordance Analysis & Collocation Analysis

Given the repeated use of the word "love" across the chart, I've decided to print my KWIC and collocation analyses of this term below.

## All 1990s

In [22]:
#create list of KWIC concordance lines of "love" from 'all_90s'
kwic_love_90s = []
for song in all_charts['all_90s']:
    if song.get('tokens'):
        kwic_rel_love90s = make_kwic('love', song['tokens'])
        kwic_love_90s.extend(kwic_rel_love90s)
        
print(f'"love" occurs {len(kwic_love_90s)} times in your lyrics')

#sort and view them
love_sorted_90s = sort_kwic(kwic_love_90s, ['L1','R1'])
print_kwic(love_sorted_90s)

"love" occurs 108 times in your lyrics
                                          love  is unconditional we knew
                     you'll never find a  love  as true as mine
                     you'll never find a  love  as true as mine
                     you'll never find a  love  as true as mine
                          true as mine a  love  as true as mine
                         and then it's a  love  without end amen it's
                         end amen it's a  love  without end amen when
                      and a little about  love  a lot about livin'
                     worn out line about  love  at first sight well
                      and a little about  love  aw haw well we
                      and a little about  love  well way down yonder
                      and a little about  love  yeaheee that's right 
                      she can't help but  love  him for the way
                         and how you can  love  like this 'cause i'm
                      

In [23]:
##love collocation 90s
love_90s_colls = Counter()
for song in range(len(all_charts['all_90s'])):
    love_90s_colls.update(collocates(tokenize(all_charts['all_90s'][song]['lyrics']),'love', [3,3]))

In [24]:
love_90s_colls.most_common(25)

[('I', 33),
 ('you', 28),
 ('the', 26),
 ('boy', 21),
 ('in', 19),
 ('(Oh,', 18),
 ('girl)', 18),
 ("It's", 15),
 ("She's", 13),
 ('with', 13),
 ('a', 13),
 ('Marie,', 12),
 ('(Maria)', 12),
 ('it,', 12),
 ('way', 11),
 ('like', 9),
 ('as', 9),
 ('your', 9),
 ('our', 8),
 ('about', 7),
 ('A', 7),
 ("Don't", 7),
 ('let', 7),
 ('start', 7),
 ("slippin'", 7)]

## All 2010s

In [25]:
#create list of KWIC concordance lines of "love" from 'all_2010s'
kwic_love_2010s = []
for song in all_charts['all_2010s']:
    if song.get('tokens'):
        kwic_rel_love2010s = make_kwic('love', song['tokens'])
        kwic_love_2010s.extend(kwic_rel_love2010s)
        
print(f'"love" occurs {len(kwic_love_2010s)} times in your lyrics')

#sort and view them
love_sorted_2010s = sort_kwic(kwic_love_2010s, ['L1','R1'])
print_kwic(love_sorted_2010s)

"love" occurs 19 times in your lyrics
                   well sunsets fade and  love  does too yeah we
                    you sunsets fade and  love  does too we had
                    you sunsets fade and  love  does too though we
                      and rescue me felt  love  pouring down from above
                  you thought you'd find  love  isn't all that it
                      someone you love i  love  you ain't no pickup
                      we were fallin' in  love  in the sweet heart
                          time i fell in  love  with a careless man's
                          all we need is  love  just as free free
                       love who you love  love  who you love 'cause
                        went up in smoke  love  isn't all that it
                    take for granted the  love  this life gives you
                      say what you think  love  who you love love
                      why we bother with  love  if it never lasts
                       

In [26]:
##love collocation 2010s
love_2010s_colls = Counter()
for song in range(len(all_charts['all_2010s'])):
    love_2010s_colls.update(collocates(tokenize(all_charts['all_2010s'][song]['lyrics']),'love', [3,3]))

In [27]:
love_2010s_colls.most_common(25)

[('you', 4),
 ('in', 3),
 ('fade,', 3),
 ('and', 3),
 ('does,', 3),
 ('too', 3),
 ('with', 2),
 ('who', 2),
 ('Sunsets', 2),
 ('the', 2),
 ('"I', 2),
 ('love', 2),
 ('you"', 2),
 ('I', 1),
 ('fell', 1),
 ('a', 1),
 ('careless', 1),
 ('Love', 1),
 ('(Love', 1),
 ('rescue', 1),
 ('me."', 1),
 ('Felt', 1),
 ('pouring', 1),
 ('down', 1),
 ('from', 1)]

### Observations - 1990s vs 2010s 

Above, I ran KWIC and collocation analyses for `all_90s` and `all_2010s` to see if the word "love" is used differently in either chart. Scanning over the printed results from the KWIC analysis, due to the repetition of specific lines, it seems that "love" may be appearing in several of the choruses of '90s songs. "Love" is either used in the lyrics as a profession of love with the pretexts of "I" (as in "I love you") or the term is used more generally, when the artists are speaking about a love story. Keeping this in mind, I predict that both the direct and indirect uses of "love" will yield similar results in the other corpora. 

Also, given that the word "love" appeared significantly less in the `all_2010s` data (19 times compared to 108 in `all_90s`), I figured the context would be different as well. With the 1990s data, "love" seemed to have primarily positive connotations; howeber, in the 2010s data, "love" seemed to occur in lines with more negative sentiments. The `all_2010s` data places "love" around words and phrases that reference love fading and love not being all it is expected to be. Even if you pay close attention to the collocations of the 2010s, "love" is surrounded by words such as "fade", "rescue", and "careless". 

All of this taken into account, there are some differences in the way the term "love" is used across decades. 

## All Female

In [28]:
#create list of KWIC concordance lines of "love" from 'all_female'
kwic_love_f = []
for song in all_charts['all_female']:
    if song.get('tokens'):
        kwic_rel_love_f = make_kwic('love', song['tokens'])
        kwic_love_f.extend(kwic_rel_love_f)
        
print(f'"love" occurs {len(kwic_love_f)} times in your lyrics')

#sort and view them
love_sorted_f = sort_kwic(kwic_love_f, ['L1','R1'])
print_kwic(love_sorted_f)

"love" occurs 49 times in your lyrics
                   well sunsets fade and  love  does too yeah we
                    you sunsets fade and  love  does too we had
                    you sunsets fade and  love  does too though we
                      she can't help but  love  him for the way
                         and how you can  love  like this 'cause i'm
                        you in a country  love  song summer night beauty
                     snow white how does  love  get so off course
                      and rescue me felt  love  pouring down from above
                  you thought you'd find  love  isn't all that it
                restless summer we found  love  growing wild on the
                      like we're even in  love  it matters to me
                      like we're even in  love  it matters to me
                 tommy anywhere she's in  love  with the boy she's
                        the boy she's in  love  with the boy she's
                        

In [29]:
##love collocation all female
love_f_colls = Counter()
for song in range(len(all_charts['all_female'])):
    love_f_colls.update(collocates(tokenize(all_charts['all_female'][song]['lyrics']),'love', [3,3]))

In [30]:
love_f_colls.most_common(25)

[('the', 23),
 ('boy', 21),
 ('in', 18),
 ('with', 14),
 ('you', 14),
 ("She's", 13),
 ('way', 11),
 ('a', 6),
 ('me', 6),
 ('(The', 6),
 ("It's", 5),
 ('me,', 5),
 ('[Chorus]', 4),
 ('fade,', 3),
 ('and', 3),
 ('does,', 3),
 ('too', 3),
 ("we're", 2),
 ('even', 2),
 ('It', 2),
 ('matters', 2),
 ('to', 2),
 ('way)', 2),
 ('baby', 2),
 ('baby)', 2)]

## All Male

In [31]:
#create list of KWIC concordance lines of "love" from 'all_male'
kwic_love_m = []
for song in all_charts['all_male']:
    if song.get('tokens'):
        kwic_rel_love_m = make_kwic('love', song['tokens'])
        kwic_love_m.extend(kwic_rel_love_m)
        
print(f'"love" occurs {len(kwic_love_m)} times in your lyrics')

#sort and view them
love_sorted_m = sort_kwic(kwic_love_m, ['L1','R1'])
print_kwic(love_sorted_m)

"love" occurs 78 times in your lyrics
                                          love  is unconditional we knew
                     you'll never find a  love  as true as mine
                     you'll never find a  love  as true as mine
                     you'll never find a  love  as true as mine
                          true as mine a  love  as true as mine
                         and then it's a  love  without end amen it's
                         end amen it's a  love  without end amen when
                      and a little about  love  a lot about livin'
                     worn out line about  love  at first sight well
                      and a little about  love  aw haw well we
                      and a little about  love  well way down yonder
                      and a little about  love  yeaheee that's right 
                   of pretending i don't  love  you anymore let me
                 keep pretending i don't  love  you anymore i've got
                   o

In [32]:
##love collocation all male
love_m_colls = Counter()
for song in range(len(all_charts['all_male'])):
    love_m_colls.update(collocates(tokenize(all_charts['all_male'][song]['lyrics']),'love', [3,3]))

In [33]:
love_m_colls.most_common(25)

[('I', 33),
 ('(Oh,', 18),
 ('you', 18),
 ('girl)', 18),
 ('Marie,', 12),
 ('(Maria)', 12),
 ('it,', 12),
 ("It's", 10),
 ('as', 10),
 ('your', 9),
 ('a', 8),
 ('our', 8),
 ('like', 8),
 ('about', 7),
 ('A', 7),
 ("Don't", 7),
 ('let', 7),
 ('start', 7),
 ("slippin'", 7),
 ('Love', 7),
 ('you,', 6),
 ('[Verse', 6),
 ('want', 6),
 ('the', 5),
 ('My', 4)]

### Observations - Female vs Male 

Analyzing the female versus male charts, three words from the `all_female` collocation analysis stood out to me: "in", "with", and "you". I want to highlight these contextual words because they  can be linked to a general message of 'being in love' or 'falling in love'. When looking at the `all_male` data, there are some differences, one main difference being that "I" prefaces "love" significantly more in songs written by male artists. Again, you can see from the concordance lines that there is a lot of repetition that stems from the word "love" appearing in core parts of the songs such as the choruses.

Thinking about these sample KWIC concordance and keyness analyses, one can determine that "love" is a unique lyrical similarity between the decades that can be used in different contexts by artists of different genders!

## Keyness Analysis

For closer analysis, I decided to run keyness analyses of the various charts to determine the most key terms in each data set. 

## 1990s vs 2010s

In [34]:
## keyness analysis: key words in the 90s subset vs those in 2010s
calculate_keyness(word_freq_90s, word_freq_2010s, top = 50)

WORD                     Corpus A Freq.Corpus B Freq.Keyness
love                     108       19        77.052
boy                      32        8         17.639
let                      47        18        16.121
tell                     36        14        12.080
want                     24        7         11.434
i've                     35        15        10.111
really                   19        5         10.003
oh                       70        42        9.794
know                     74        46        9.329
say                      33        15        8.618
maybe                    19        6         8.312
said                     32        16        6.959
would                    22        9         6.844


## 2010s vs 1990s

In [35]:
## keyness analysis: key words in the 2010s subset vs those in 90s
calculate_keyness(word_freq_2010s, word_freq_90s, top = 50)

WORD                     Corpus A Freq.Corpus B Freq.Keyness
back                     80        15        43.221
every                    51        10        26.601
'em                      39        6         24.190
road                     34        5         21.687
like                     108       55        13.149
need                     33        10        11.004
good                     33        12        8.408
free                     22        6         8.349
always                   19        5         7.507
whiskey                  20        6         6.757


### Observations:

- No clear plot can be distinguished from either of these keyness charts 
- "Love" appears to be significantly more key in the `all_90s` chart than in `all_2010s`; this could be because "love" is repeatedly used in core parts of the song such as the chorus
- "Boy" is another key word than appears and may have more contextual, lyrical variety than "love"
- "Really" is the only evaluative word in the first chart
- "Like" is more present in country songs of the 2010s than the 1990s; can be used in a variety of ways 
- "Whiskey" and "free" are more key in 2010s than in 1990s

## Female vs Male

In [36]:
calculate_keyness(f_word_freq, m_word_freq, top = 50)

WORD                     Corpus A Freq.Corpus B Freq.Keyness
boy                      34        6         23.204
every                    46        15        18.263
kiss                     27        9         10.435
said                     34        14        9.728
prechorus                24        8         9.275
thing                    25        9         8.740
ever                     37        17        8.726
let                      43        22        8.111
name                     22        8         7.586
without                  18        6         6.956


## Male vs Female

In [37]:
calculate_keyness(m_word_freq, f_word_freq, top = 50)

WORD                     Corpus A Freq.Corpus B Freq.Keyness
wanna                    57        14        25.619
i'm                      114       58        15.622
girl                     40        17        8.330


### Observations:

- I find it interesting that "boy" was key term in the `all_female` chart and "girl" was significant in the `all_male` chart
- Some reference to the theme of 'love' with the term "kiss" in the `all_female` chart
- "Wanna" and "I'm" don't leave much room for analysis from the Male vs Female Keyness

## Part of Speech Tagging

For the last part of my data story, I wanted to share my part of speech analysis that aimed to determined what kinds of words were used most frequently within the different decades and by artists of different genders.

## All 1990s

In [44]:
## what percentage of all tokens in 90s corpus are verbs/adjectives
## divide sum of each counter by sum of total tokens counter

verb_90s_perc = (sum(verb_90s_freq.values()) / sum(word_freq_90s.values()))* 100
adj_90s_perc = (sum(adj_90s_freq.values()) / sum(word_freq_90s.values()))* 100

print('{} percent of the tokens in all 90s songs were verbs'.format(verb_90s_perc))
print('{} percent of the tokens in all 90s songs were adjectives'.format(adj_90s_perc))

42.479946524064175 percent of the tokens in all 90s songs were verbs
17.34625668449198 percent of the tokens in all 90s songs were adjectives


## All 2010s

In [43]:
# what percentage of all tokens in 2010s corpus are verbs/adjectives
## divide sum of each counter by sum of total tokens counter

verb_2010s_perc = (sum(verb_2010s_freq.values()) / sum(word_freq_2010s.values()))* 100
adj_2010s_perc = (sum(adj_2010s_freq.values()) / sum(word_freq_2010s.values()))* 100

print('{} percent of the tokens in all 2010s songs were verbs'.format(verb_2010s_perc ))
print('{} percent of the tokens in all 2010s songs were adjectives'.format(adj_2010s_perc))

39.8196270253745 percent of the tokens in all 2010s songs were verbs
16.70742892081932 percent of the tokens in all 2010s songs were adjectives


### Observations:

Looking at these results, the proportion of verbs and adjectives in the 1990s and 2010s charts seem to be nearly identical, revealing that not one decade had more action words or descriptive words than the other. 

Now, I will take a look at these percentages within the charts of songs written my male and female artists.

## All Female

In [45]:
verb_female_perc = (sum(verb_female_freq.values()) / sum(f_word_freq.values()))* 100
adj_female_perc = (sum(adj_female_freq.values()) / sum(f_word_freq.values()))* 100

print('{} percent of the tokens in all female songs were verbs'.format(verb_female_perc))
print('{} percent of the tokens in all female songs were adjectives'.format(adj_female_perc))

43.145624692168774 percent of the tokens in all female songs were verbs
17.271384009193895 percent of the tokens in all female songs were adjectives


## All Male

In [47]:
verb_male_perc = (sum(verb_male_freq.values()) / sum(m_word_freq.values()))* 100
adj_male_perc = (sum(adj_male_freq.values()) / sum(m_word_freq.values()))* 100

print('{} percent of the tokens in all male songs were verbs'.format(verb_male_perc))
print('{} percent of the tokens in all male songs were adjectives'.format(adj_male_perc))

39.14529914529915 percent of the tokens in all male songs were verbs
16.767676767676768 percent of the tokens in all male songs were adjectives


### Observations: 

Looking above, female and male artists seem to use verbs and adjectives in almost equal proportions as well, showing that the part of speeches of words across time are pretty consistent. Additionally, verbs are used significantly more than adjectives by both male and female artists. 

In fact, within the part of speech tagging referenced in this section, all of the same verbs and adjectives appeared, which references the remarkable similarity in country music over time. 

## CONCLUDING REMARKS / DISCUSSION

While working on this final project, one of the primary questions and worries I had with this endeavor was, even if particular words were used across decades and genders, how would I really be able to distinguish the true contexts of each word being used? More specifically, how could I distinguish musical sound-based differences from textual, lyrical differences? 

Before delving into my findings, I think it is important to acknowledge inherent drawbacks of my project before I highlight the strengths! Given the specificity of my research questions -  personally choosing which decades to analyze and which songs to filter through may have skewed my analysis. I believe that including hit songs from country's earliest days may have been a really interesting additional component to my research project. That being said, I still think I discovered some valuable insight on how the country music genre has evolved over time. 

Going through this notebook, I highlighted some lyrical similarities and differences found in lyrics between the decades and by male and female artists. I found that words and phrases used in the 1990s were also used more recently in country music of the 2010s. While shifts in country music sound and style may have surprised country genre supporters and artists, it seems to be that the messages and preaches of songs over time haven't changed all that much. 

When looking at similarities and differences between male and female artists, my hypothesis standards correct - that there *are in fact*  more significant distinctions between lyrics written by male and female artists than there are between lyrics of different times. 

If I were to build on my current analysis for this project, I would have taken a closer look at the actual songwriters - not just the artist(s) of every song - because that might reveal even more particular results than what I explored throughout this notebook and my data analysis notebooks. In other words, my analysis might have been more targeted and precise if I analyzed songs written by the best songwriters of specific decades and then split them up into gender dictionaries. 

Thank you for coming along with me on my journey to understanding the evolution of country music over time! I hope you enjoyed it!